File: app/src/main/java/chess/android/arduino/telescope/coordinates/Coordinate.java
package chess.android.arduino.telescope.coordinates;

import java.time.LocalTime;

public abstract class Coordinate {
    private int degree;
    private LocalTime timer;

    Coordinate(int hour, int minute, int second) {
        if (hour>24 || hour<0 || minute>59 || minute<0 || second>59 || second<0) {
            throw new IllegalArgumentException(String.format("Illegal time: %d, %d, %d", hour, minute, second));
        }
        timer = LocalTime.of(hour, minute, second);
    }

    public int getDegree() {
        return degree;
    }

    void setDegree(int degree) {
        this.degree = degree;
    }

    public int getHour() {
        return  this.timer.getHour();
    }

    public int getMinute() {
        return this.timer.getMinute();
    }


    public int getSecond() {
        return this.timer.getSecond();
    }

    public void plusHour() {
        timer = timer.plusHours(1);
    }

    public void plusDegree() {
        if(degree!=89) degree++;
    }

    public void minusDegree() {
        if(degree!=-89) degree--;
    }


    public void minusHour() {
        timer = timer.plusHours(-1);
    }

    public void plusMinute() {
        timer = timer.plusMinutes(1);
    }

    public void minusMinute() {
        timer = timer.plusMinutes(-1);
    }

    public void  plusSecond() {
        timer = timer.plusSeconds(1);
    }

    public void minusSecond() {
        timer = timer.plusSeconds(-1);
    }

    public abstract double toDegrees();




}

File: app/src/main/java/chess/android/arduino/telescope/coordinates/Coordinate.kt
package chess.android.arduino.telescope.coordinates

import java.time.LocalTime

abstract class Coordinate internal constructor(hour: Int, minute: Int, second: Int) {
    var degree = 0
    private var timer: LocalTime

    val hour: Int
        get() = timer.hour

    val minute: Int
        get() = timer.minute

    val second: Int
        get() = timer.second

    fun plusHour() {
        timer = timer.plusHours(1)
    }

    fun plusDegree() {
        if (degree != 89) degree++
    }

    fun minusDegree() {
        if (degree != -89) degree--
    }

    fun minusHour() {
        timer = timer.plusHours(-1)
    }

    open fun plusMinute() {
        timer = timer.plusMinutes(1)
    }

    open fun minusMinute() {
        timer = timer.plusMinutes(-1)
    }

    fun plusSecond() {
        timer = timer.plusSeconds(1)
    }

    fun minusSecond() {
        timer = timer.plusSeconds(-1)
    }

    abstract fun toDegrees(): Double

    init {
        require(!(hour > 24 || hour < 0 || minute > 59 || minute < 0 || second > 59 || second < 0)) { String.format("Illegal time: %d, %d, %d", hour, minute, second) }
        timer = LocalTime.of(hour, minute, second)
    }
}